odl/space/base_tensors.py

```diff
@@ -0,0 +1,741 @@
+        method.
+        """
+        return type(self)(self.shape, dtype=dtype, order=order,
+                          weighting=getattr(self, 'weighting', None))
+
+    def astype(self, dtype, order=None):
+        """Return a copy of this space with new ``dtype``.
+
+        Parameters
+        ----------
+        dtype :
+            Scalar data type of the returned space. Can be provided
+            in any way the `numpy.dtype` constructor understands, e.g.
+            as built-in type or as a string. Data types with non-trivial
+            shapes are not allowed.
+        order : {'A', 'C', 'F'}, optional
+            Axis ordering of the data storage. Only relevant for more
+            than 1 axis.
+            For ``'C'`` and ``'F'``, elements are forced to use
+            contiguous memory in the respective ordering.
+            For ``'A'`` ("any") no contiguousness is enforced.
+            The default ``None`` is equivalent to ``self.order``.
+
+        Returns
+        -------
+        newspace : `TensorSpace`
+            Version of this space with given data type.
+        """
+        if dtype is None:
+            # Need to filter this out since Numpy iterprets it as 'float'
+            raise ValueError('`None` is not a valid data type')
+
+        if order is None:
+            order = self.order
+
+        dtype = np.dtype(dtype)
+        if dtype == self.dtype and order == self.order:
+            return self
+
+        if self.is_numeric:
+            # Caching for real and complex versions (exact dtype mappings)
+            if dtype == self.__real_dtype and order == self.order:
+                if self.__real_space is None:
+                    self.__real_space = self._astype(dtype, self.order)
+                return self.__real_space
+            elif dtype == self.__complex_dtype and order == self.order:
+                if self.__complex_space is None:
+                    self.__complex_space = self._astype(dtype, self.order)
+                return self.__complex_space
+            else:
+                return self._astype(dtype, order)
+        else:
+            return self._astype(dtype, order)
+
+    @property
+    def order(self):
+        """Guaranteed data storage order in this space.
+
+        This is one of ``('A', 'C', 'F')``, where ``'A'`` means "any",
+        i.e. "no guarantee".
+        """
+        return self.__order
+
+    @property
+    def new_elem_order(self):
+        """Storage order for new elements in this space.
+
+        This is identical to `order` except for ``self.order == 'A'``,
+        where ``'C'`` is returned.
+        """
+        return 'C' if self.order == 'A' else self.order
+
+    @property
+    def size(self):
+        """Total number of entries in an element of this space."""
+        return np.prod(self.shape)
+
+    @property
+    def ndim(self):
+        """Number of axes (=dimensions) of this space, also called "rank"."""
```

Comment:
Well at least in tensor lingo, number of dimensions and rank are the same thing. However, when the dimensions have different interpretations, it sure needs additional thought. Most important is to keep compatibility with the `ProductSpace` version of the same thing. Luckily, those don't have a shape yet, so there's no ambiguity at the moment.

New issue I guess, and I remove the rank remark here.